File: 2019/day15/cpp/src/droid.cpp
#include "droid.hpp"

#include <signal.h>

#include <queue>

#include "ncurses.h"

Droid::Droid(std::string file_name) : m_intcode(), m_map() {
  m_intcode = Intcode(file_name, {}, false, false);
}

// Return new output pair depdning on input
std::pair<int, int> Droid::one_step(int y, int x, int input) {
  y -= (input == 1) ? 1 : ((input == 2) ? -1 : 0);
  x += (input == 3) ? 1 : ((input == 4) ? -1 : 0);
  return std::make_pair(y, x);
}

std::tuple<int, int, int> Droid::repair_oxygen(int y_, int x_) {
  // Initialize starting position
  std::map<std::pair<int, int>, int> parents;
  parents[std::make_pair(y_, x_)] = 1;
  m_map[std::make_pair(y_, x_)] = 1;

  // Use BFS to explore the "map"
  std::queue<std::tuple<int, int, Intcode, int>> q;
  q.push(std::make_tuple(y_, x_, m_intcode, 0));

  // Output solution
  std::tuple<int, int, int> solution;

  // Using ncurses
  signal(SIGINT, catch_ctrlc);
  initscr();

  // Run until we have explored everything
  while (!q.empty()) {
    // Get a queue_data struct
    auto q_ = q.front();
    q.pop();
    auto y = std::get<0>(q_);
    auto x = std::get<1>(q_);
    auto ic = std::get<2>(q_);
    auto depth = std::get<3>(q_);

    for (int i = 0; i < 5; i++) {
      auto xy_new = one_step(y, x, i);
      auto y_new = xy_new.first;
      auto x_new = xy_new.second;

      // If we havent visited the state
      if (parents.count(xy_new) != 1) {
        parents[xy_new] = 1;

        Intcode ic_ = ic;
        auto out = std::get<1>(ic_(i));

        if (out == 1) {
          m_map[xy_new] = 1;
          q.push(std::make_tuple(y_new, x_new, ic_, depth + 1));
        } else if (out == 2) {
          m_map[xy_new] = 2;
          solution = std::make_tuple(y_new, x_new, depth + 1);
        } else {
          m_map[xy_new] = 0;
        }
      }
    }
    render();
  }
  endwin();
  return solution;
}

int Droid::fill_oxygen(int y_, int x_) {
  // Initialize starting position
  std::map<std::pair<int, int>, int> parents;
  parents[std::make_pair(y_, x_)] = 1;

  // BFS search in the map
  std::queue<std::tuple<int, int, int>> q;
  q.push(std::make_tuple(y_, x_, 0));

  signal(SIGINT, catch_ctrlc);
  initscr();

  // Keep going through all empty map spaces and keep count
  auto max_depth = -1;
  while (!q.empty()) {
    auto q_ = q.front();
    q.pop();
    auto y = std::get<0>(q_);
    auto x = std::get<1>(q_);
    auto depth = std::get<2>(q_);

    for (int i = 0; i < 5; i++) {
      auto xy_new = one_step(y, x, i);
      auto y_new = xy_new.first;
      auto x_new = xy_new.second;

      if (parents[xy_new] != 1) {
        parents[xy_new] = 1;

        if (m_map[xy_new] == 1) {
          m_map[xy_new] = 2;
          max_depth = depth + 1;
          q.push(std::make_tuple(y_new, x_new, depth + 1));
        }
      }
    }
    render();
  }
  endwin();
  return max_depth;
}

// Print the map
void Droid::render() {
  int height, width;
  getmaxyx(stdscr, height, width);

  int max_x(WINT_MIN), min_x(WINT_MAX);
  int max_y(WINT_MIN), min_y(WINT_MAX);

  for (auto& m : m_map) {
    max_x = (max_x < m.first.first) ? m.first.first : max_x;
    max_y = (max_y < m.first.second) ? m.first.second : max_y;

    min_x = (min_x > m.first.first) ? m.first.first : min_x;
    min_y = (min_y > m.first.second) ? m.first.second : min_y;
  }

  for (int row = min_x; row < max_x + 1; row++) {
    std::string str = "";
    for (int col = min_y; col < max_y + 1; col++) {
      int mp = m_map[std::pair<int, int>(row, col)];
      str += (mp == 0) ? "X" : ((mp == 2) ? "O" : " ");
      if (col > width - 3) {
        break;
      }
      mvprintw(row - min_x, 0, str.c_str());
    }
    if (row - min_x > height - 3) {
      break;
    }
  }
  refresh();
}

void Droid::catch_ctrlc(int s) {
  endwin(); /* End curses mode		  */
  printf("Caught signal %d\n", s);
  exit(1);
}

File: 2019/day15/python/droid.py
from os.path import dirname
from os.path import realpath
from os.path import join
from collections import defaultdict
from queue import Queue
import sys
import copy
import curses

sys.path.append(join(dirname(realpath(__file__)), *["..", ".."]))

from day5.python.intcode import Intcode


class Droid:
    """Class that controls the droid."""

    def __init__(self, fileName):
        """Initialize class and intcode computer."""
        self.ic = Intcode(fileName, verbose=False, reset=False)

    def one_step(self, y, x, inp):
        """Returns next (y,x) pair given input."""
        y -= 1 if inp == 1 else -1 if inp == 2 else 0
        x += 1 if inp == 3 else -1 if inp == 4 else 0
        return y, x

    def repair_oxygen(self, y=0, x=0):
        """Starts from (y,x) and searches for the oxygen (output 2 from intcode).
        Program continues to map entire area after finding the oxygen."""

        # Keep track of parent nodes
        parents = defaultdict(list)
        parents[y, x] = (-1, -1)

        # Create a map of the environment
        map = defaultdict(lambda: 0)
        map[y, x] = 1

        # Queue - first come first serve
        q = Queue()
        q.put((y, x, copy.deepcopy(self.ic), 0))

        output = None

        # Initialize screen
        screen = curses.initscr()
        curses.cbreak()

        try:
            while not q.empty():
                y, x, ic_, depth = q.get()
                for input in range(1, 5):
                    y_new, x_new = self.one_step(y, x, input)

                    if not len(parents[y_new, x_new]) > 0:
                        parents[y_new, x_new] = (y, x)

                        ic_new = copy.deepcopy(ic_)
                        _, out = ic_new(input)
                        if out == 1:
                            map[y_new, x_new] = 1  # Plotting purpose
                            q.put((y_new, x_new, ic_new, depth + 1))
                        elif out == 2:
                            map[y_new, x_new] = 2  # Plotting purpose
                            output = (y_new, x_new, depth + 1)
                        else:
                            map[y_new, x_new] = 0  # Plotting purpose

                self.render(screen, map)  # Plot map
        finally:
            curses.echo()
            curses.nocbreak()
            curses.endwin()
            return map, output

    def render(self, screen, mp):
        """Go through the game map and print the walls/obstacles/pad/ball."""
        try:
            height, width = screen.getmaxyx()

            y_max, x_max = map(max, zip(*mp.keys()))
            y_min, x_min = map(min, zip(*mp.keys()))

            for row, v in enumerate(range(y_min, y_max + 1)):
                msg = []
                for k in range(x_min, x_max + 1):
                    if mp[v, k] == 0:
                        msg.append("X")
                    elif mp[v, k] == 1:
                        msg.append(" ")
                    elif mp[v, k] == 2:
                        msg.append("O")
                    else:
                        msg.append(" ")
                    if k > width - 3:
                        break
                screen.addstr(row, 0, "".join(msg))
                if row > height - 3:
                    break
            screen.refresh()
        except KeyboardInterrupt:
            curses.echo()
            curses.nocbreak()
            curses.endwin()
            exit(0)

    def fill_oxygen(self, map, y=0, x=0):
        """Given a starting position, sees how many time steps (how deep) the
        search goes until entire map is explored."""

        # Keep track of parent nodes
        parents = defaultdict(list)
        parents[y, x] = (-1, -1)

        # Queue - first come first serve
        q = Queue()
        q.put((y, x, 0))

        max_depth = -1
        screen = curses.initscr()
        curses.cbreak()
        try:
            while not q.empty():
                y, x, depth = q.get()
                for input in range(1, 5):

                    y_new, x_new = self.one_step(y, x, input)

                    if not len(parents[y_new, x_new]) > 0:

                        parents[y_new, x_new] = (y, x)
                        if map[y_new, x_new] == 1:
                            map[y_new, x_new] = 2
                            max_depth = depth + 1
                            q.put((y_new, x_new, depth + 1))
                self.render(screen, map)

        finally:
            curses.echo()
            curses.nocbreak()
            curses.endwin()
            return max_depth
